### DISCUSSION TITLE
Upgrade from previous hardware

### DISCUSSION THREAD

**Discussion by JeanBaptiste-elec** [Arnold-n/P1P2MQTT#71] [Q&A]
Hello,

I've been using the P1P2serial for several years. I use the previous hardware : P1P2-adapter on Arduino Uno.
I would like to upgrade to the last version of the software, but the support of MQTT was removed in P1P2serial.
Is it possible to add some hardware in order to convert the serial informations to MQTT protocol ?
I Saw that there is some card with ATMega328 and ESP8266 on the same PCB. Can I use this ?
I would prefer to use an ethernet connection.

Or, is it necessary to get the last hardware revision ?
Thanks for your very useful work.

**Comment by Arnold-n** [comment]
Hello @JeanBaptiste-elec,

Support for MQTT was removed from the Uno due to the limited size of the ATmega flash memory. You can indeed add it by adding an ESP, but:
-the Uno is 16MHz/5V while the ATmega in the new P1P2MQTT hardware runs at 8MHz/3V3, and the new library ("NEWP1P2LIB") is written and tested for 8MHz; until P1P2Serial v0.9.18 the OLDP1P2LIB option supported 16MHz I plan to reintroduce that option,
-you can run the MQTT code separately on an ESP but the ESP is 3.3V and the Uno is 5V so the serial connection must have a level-shifter,
-the ESP has only WiFi  and requires an extra w5500 adapter to add ethernet support, and
-the ATmega328/ESP8266 combi-board you mention is very convenient and does the level-shifting but the manufacturer went out of business.

The new hardware is more convenient in use as it is a small PCB and bus-powered and fits nicely in a 40x80x24mm enclosure (or 40x80x20 without ethernet). I'll mail you directly on the new hardware version.

**Reply by JeanBaptiste-elec** [reply]
Hello,
Thank you for your reply. Current options are clear.
I bouhgt yesterday the combiboard (or a clone?) on Ali... 
I don't know if its pinout is compatible with P1P2-adapter. It will be a surprise.
I gess that I should use Dupont connectors in order to connect a W5500 adapter or the ESP header can be use ?

**Reply by Arnold-n** [reply]
On the original combi-board the Uno connectors are likely the same as those on the original Uno. Connecting the W5500 to the ESP may be a challenge: this requires connecting GPIO4/16/12/13/14 to SCS/RST/MISO/MOSI/CLK, and I am not sure all of these are easily available for connection on the combi-board: on my combi-board GPIO4/16/12 are available through the male header pins but GPIO13/14 are not easily available. It may be easier to use a separate ESP12F and a level-shifter, or stay with WiFi.
